### PR TITLE
Issue 6

### LINKED ISSUE #6: support environment variable configure source
need supporting parse configure from environment variables.
Start in go file: json_func.go
everything you should know about this case - in your configure struct. 
This case should be contain methods which can parse env like default format and tag name format. This case also should include variant for insert validation middleware like this:
```golang
type ExampleStruct struct {
     Email strnig `cf_env:"validate:email"`
}
```

And in tag cf_env can be name of env variable path and any other settings, for example:
```golang
type ExampleStruct struct {
     Email strnig `cf_env:"path:test_field,validate:email"`
}
In this case you should parse path and add after configure defer function Validate
```
Also in tag cf_env can be only path, and in this case you should parse only path:
```golang
type ExampleStruct struct {
     Email strnig `cf_env:"test_field_name"`
}
```

THIS ISSUE SHOULD RUN EXECUTING AFTER SPECIFIC MIDDLEWARE FUNCTIONS INTERFACE. CHECK ISSUE #6

Need support all of this types: int8, int16, int32, int64, float16, float32, float64, string, bool, char, slice, map

### PR DESCRIPTION
**Pull request by kubitre**
resolve #6

### CODE REVIEW

**Comment by kubitre** [comment]
resolve #21

File: pipeline/environment_function_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+				},
+			},
+			want: infra.NewGoStructorTrueValue(reflect.ValueOf("tururu")),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			config := EnvironmentConfig{}
+			got := config.GetBaseType(tt.args.context)
+			os.Remove("testBaseType")
```
Comment (artemkaxboy) [inline pipeline/environment_function_test.go]: Don't do this in a loop. Use defer function right after os.Setenv

File: pipeline/pipeline.go [hunk 2]
```diff
@@ -36,21 +33,17 @@ type (
...
+	/*DirtyConfiguring - flag*/
+	DirtyConfiguring = false
 
 	sourceFileInDisk   = 0
 	sourceFielInServer = 1
 	sourceFileNotUsed  = 2
 
 	notSupportedTypeError = "not supported type "
 
+	/*EmptyAdditionalPrefix - prefix for setup before all values can be empty*/
```
Comment (artemkaxboy) [inline pipeline/pipeline.go]: prefix for setting up / prefix to setup
And not clear what you mean here, maybe you missed a comma after values or i didn't get it?

File: pipeline/pipeline.go [hunk 3]
```diff
@@ -159,7 +154,15 @@ func Configure(
...
 	}()
 
-	pipeline := getFunctionChain(fileName, pipelineChaines)
+	var pipeline *Pipeline
+
+	if smartConfigure {
+		analysedChains := tags.GetFunctionTypes(structure)
+		pipeline = getFunctionChain(fileName, analysedChains)
+	} else {
+		pipeline = getFunctionChain(fileName, pipelineChaines)
```
Comment (artemkaxboy) [inline pipeline/pipeline.go]: pipelineChai**ns**

File: tags/analys_func.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
+package tags
+
+import (
+	"reflect"
+
+	"github.com/goreflect/gostructor/infra"
+)
+
+/*
+GetFunctionTypes - return slice of functions which should configuring sourceStruct structure
```
Comment (artemkaxboy) [inline tags/analys_func.go]: should configur**e**

File: tags/analys_func.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+*/
+func GetFunctionTypes(sourceStruct interface{}) []infra.FuncType {
+	summirize := []int{}
```
Comment (artemkaxboy) [inline tags/analys_func.go]: I believe it must be summ**a**rize

File: tags/analys_func.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	value := reflect.Indirect(reflect.ValueOf(sourceStruct))
+	for i := 0; i < value.NumField(); i++ {
+		summirizeLevel := recurseStructField(value.Type().Field(i))
```
Comment (artemkaxboy) [inline tags/analys_func.go]: And summ**a**rizeLevel

File: tags/analys_func.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	summirize := checkFuncsByTags(structField)
+
+	switch structField.Type.Kind() {
+	case reflect.Struct:
+		for i := 0; i < structField.Type.NumField(); i++ {
+			summirizeLevel := recurseStructField(structField.Type.Field(i))
+			summirize = combineFields(summirize, summirizeLevel)
+		}
+	}
+	return summirize
```
Comment (artemkaxboy) [inline tags/analys_func.go]: summ**a**rize

File: tags/analys_func.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		tagInField := structField.Tag.Get(value)
+		if tagInField == "" {
+			continue
+		} else {
+			// TODO: add additional anaylys tag values for middlewares functions and others
+			summirize[getFuncTypeByTag(value)]++
+		}
+	}
+
+	return summirize
```
Comment (artemkaxboy) [inline tags/analys_func.go]: please check the whole project summ**a**rize

File: lib.go [hunk 9]
```diff
@@ -24,6 +25,16 @@ func ConfigureSetup(
...
 	fileName string,
 	prefix string,
-	functions []pipeline.FuncType) (interface{}, error) {
-	return pipeline.Configure(structure, fileName, functions, prefix)
+	functions []infra.FuncType) (interface{}, error) {
+	return pipeline.Configure(structure, fileName, functions, prefix, pipeline.DirtyConfiguring)
+}
+
+/*
+ConfigureSmart - configuring by analys tags for add prefer strategy for configuring
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib.go:33]: ++

File: lib_test.go [hunk 10]
```diff
@@ -91,3 +112,29 @@ func Test_parseHoconWithNodeNotation2(t *testing.T) {
...
+		Field2: "test",
+		Field3: true,
+		Field4: 12.2,
+		Field5: []bool{true, false, true},
+	}, myStruct.(*EnvStruct))
+	os.Remove("myField1")
+	os.Remove("myField2")
+	os.Remove("myField3")
+	os.Remove("myField4")
+	os.Remove("myField5")
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:139]: ++

File: pipeline/environment_function.go [hunk 11]
```diff
@@ -3,21 +3,65 @@ package pipeline
...
+	switch valueIndirect.Kind() {
+	case reflect.Slice:
+		valueTag := context.StructField.Tag.Get(tags.TagEnvironment)
+		if valueTag != "" {
+			value := os.Getenv(valueTag)
+			// add here additional logic for middlewares and other
+			array := config.convertStringIntoArray(value)
+			return converters.ConvertBetweenComplexTypes(reflect.ValueOf(array), valueIndirect)
+		}
+		return infra.NewGoStructorNoValue(context.Value.Interface(), errors.New("complex type "+valueIndirect.Kind().String()+" not implement in environment parsing function"))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply pipeline/environment_function.go:37]: ++

File: pipeline/environment_function.go [hunk 12]
```diff
@@ -3,21 +3,65 @@ package pipeline
...
+	default:
+		return infra.NewGoStructorNoValue(context.Value.Interface(), errors.New("complex type "+valueIndirect.Kind().String()+" not implement in environment parsing function"))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply pipeline/environment_function.go:39]: ++

File: pipeline/environment_function.go [hunk 13]
```diff
@@ -3,21 +3,65 @@ package pipeline
...
+	fmt.Println("Level: Debug. Environment values sources start")
+	valueIndirect := reflect.Indirect(context.Value)
+	valueTag := context.StructField.Tag.Get(tags.TagEnvironment)
 
-	return infra.NewGoStructorNoValue(context.Value.Interface(), errors.New("getBaseType not implemented for environment configuring"))
+	if valueTag != "" {
+		value := os.Getenv(valueTag)
+		return converters.ConvertBetweenPrimitiveTypes(reflect.ValueOf(value), valueIndirect)
+	}
+	return infra.NewGoStructorNoValue(context.Value, errors.New("getBaseType can not getting field by empty tag value of tag: "+tags.TagEnvironment))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply pipeline/environment_function.go:60]: ++

File: tags/analys_func.go (hunk 4 above)
Comment (kubitre) [reply tags/analys_func.go:10]: ++

File: tags/analys_func.go (hunk 5 above)
Comment (kubitre) [reply tags/analys_func.go:13]: ++

File: tags/analys_func.go (hunk 6 above)
Comment (kubitre) [reply tags/analys_func.go:16]: ++

File: tags/analys_func.go (hunk 7 above)
Comment (kubitre) [reply tags/analys_func.go:38]: ++

File: tags/analys_func.go (hunk 8 above)
Comment (kubitre) [reply tags/analys_func.go:79]: ++

File: pipeline/environment_function_test.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+func TestEnvironmentConfig_GetBaseType(t *testing.T) {
+	myStruct := struct {
+		field1 string `cf_env:"testBaseType"`
+	}{}
+	os.Setenv("testBaseType", "tururu")
+	myStruct1 := reflect.Indirect(reflect.ValueOf(myStruct))
+	field1 := myStruct1.Type().Field(0)
+	t.Log("type of field: ", field1.Name+", "+field1.Type.Name())
```
[the remark this answers is not in the document]
Comment (kubitre) [reply pipeline/environment_function_test.go:20]: the code didn't get any worse

[the remark this answers is not in the document]
Comment (kubitre) [reply pipeline/environment_function_test.go:20]: moreover these are just tests

File: pipeline/environment_function_test.go (hunk 1 above)
Comment (kubitre) [reply pipeline/environment_function_test.go:46]: in our case it is not be a loop it's test contain only one case

File: pipeline/pipeline.go (hunk 2 above)
Comment (kubitre) [reply pipeline/pipeline.go]: i remove this comment

File: pipeline/pipeline.go (hunk 3 above)
Comment (kubitre) [reply pipeline/pipeline.go]: ++

**Pull request merged**